[Doc Improvement][Teams-SSO-tab] (#5567)
MicrosoftDocs/msteams-docs: msteams-platform/tabs/how-to/authentication/tab-sso-overview.md

Here's what you'll learn in this section:

1. **SSO user experience**: Teams offers your app users a true SSO experience. App users can use your app without signing in again.
2. **SSO in Teams at runtime**: Your tab app interacts with Azure AD at runtime for one-time authenticating and authorizing of your app users.
3. **Enable SSO for your tab app**: Implement the tasks involved to implement SSO in your tab app.

## SSO user experience in Teams

App users sign in to Teams using either personal Microsoft account or Microsoft 365 account. You can take advantage of this, and use SSO to authenticate and authorize the app users.

&nbsp;&nbsp;&nbsp;&nbsp; :::image type="content" source="../../../assets/images/authentication/teams-sso-tabs/teams-sso-ux.png" alt-text="SSO user experience in a Teams tab app" border="false":::

- Teams authenticates and stores the identity of its app user.
- Your tab app uses the stored identity of the app user who is already validated by Teams.

> - The app user needs to give consent to Teams for using the identity to access for using your tab app.

Reviewer: The app user or a tenant administrator can give consent. The administrator can consent for all users, removing the need to prompt the user.

Reviewer: Some permissions require an admin to consent in fact - users can't grant permission to very powerful permission scopes such as Sites.ReadWrite.All (allows reading and writing all SharePoint and OneDrive assets in the whole tenant).

- After consenting to Teams the first time, the app user can use your tab app with no further need of consent, even on any other device. For this reason, it's true SSO and offers a better user experience.
- The access token is pre-fetched by Teams to improve performance and load time of the app in Teams environment.
- App users don't need to memorize or record several passwords to access and use apps in Teams environment.

Now, let's see what happens at the backend during runtime to achieve true SSO user experience within Teams.

## SSO in Teams at runtime

Achieve SSO in a tab app by obtaining access token for the Teams app user who's currently logged in. This process involves the tab app, Teams, and Azure AD. During this interaction, the app user must give consent for using identity to obtain the access token.

The following image shows how SSO works when a Teams app user attempts to access the tab app:

:::image type="content" source="../../../assets/images/authentication/teams-sso-tabs/sso-runtime-seqd.png" alt-text="Tab single sign-on SSO diagram" border="false" lightbox="../../../assets/images/authentication/teams-sso-tabs/sso-runtime-seqd.png":::

<!--
```mermaid

> sequenceDiagram

Reviewer: SO I think the diagram should show steps  #1, then #3, 4 and 5 and then the Tab app is checking is requesting another access code and then checking if consent is needed and prompting for it.

> sequenceDiagram

Reviewer: Again TeamsFx may change the way this works so apologies if that's the case - in that case the text should specifically point out that getting the Teams client to check for and handle consent requires a certain SDK etc.
Thanks!

| 3 | Teams Client -> Azure AD | Teams requests Azure AD endpoint for the access token for the current app user based on Teams identity. |
| 4 | Azure AD -> Teams Client | Azure AD sends the access token to the Teams Client. The token is a JSON Web Token (JWT), and it's validation works just like token validation in most standard OAuth flows. Teams caches the token on your behalf so that future calls to `getAuthToken()` return the cached token. |
| 5 | Teams Client -> Tab app | Teams sends the access token to the tab app as part of the result object returned by the `getAuthToken()` call. |
| 6 | Tab app | The tab app parses the access token using JavaScript to extract required information, such as the app user's email address. The token returned to the tab app is both an access token and an identity token. |

> [!IMPORTANT]

> > The `getAuthToken()` is valid only for consenting to a limited set of user-level APIs, such as email, profile, offline_access, and OpenId. It isn't used for other Graph scopes such as `User.Read` or `Mail.Read`. For suggested workarounds, see [Extend your app with Microsoft Graph permissions](tab-sso-graph-api.md).

Reviewer: This is true - and it's also notable that this token is an access token for calling your own back-end API. (ie. the audience claim is set to your app). This allows your app to authenticate the user and authorize access to its own API. If the application doesn't need the Microsoft Graph or other services, that's all that's needed.
Calling the Graph is indeed another topic!

Tabs are Teams-aware web pages. To enable SSO in a web-page hosted inside a tab app, add [Teams Javascript client SDK](/javascript/api/overview/msteams-client?view=msteams-client-js-latest&preserve-view=true), and call `microsoftTeams.initialize()`. After you've initialized Microsoft Teams, call `microsoftTeams.getAuthToken()` to get the access token for your app.

### Use cases for enabling SSO

You can enable SSO in Teams for all apps that support Azure AD as an identity provider. In addition to using SSO for authenticating app users in a tab app, you can also use it to enable seamless access across Teams. Some scenarios where you can use `getAuthToken()` to use Teams ID token for allowing access to your app users:

- If your app already exists and you want it to be available in Teams within a tab app, you can use SSO to allow your users access to your app from within the tab app. Based on the app user's identity within Teams, you can obtain access token for them using `getAuthToken()`.

> - If your app has configurable tabs, you can use `getAuthToken()` to allow app users into the tab app without signing in again. SSO is applicable to some settings that need to be configured at a configuration stage.

Reviewer: Would it be clearer to say:
"you can use getAuthToken() in your configuration page"

This also works in configuration pages for messaging extensions and connectors.

> - Your app may have a bot, a tab, a message extension, or Adaptive Cards, and allows your app users to invoke task module from within a bot or a tab. Enabling SSO and using `getAuthToken()` will let your app users access task module from within a bot app without having to sign in again. Teams can use the app user's Teams identity and obtain an access token from Azure AD.

Reviewer: typo? Needs an article or pronoun before "task module" such as "...will let app user access *a* task module..."

- You can also use SSO to allow your app users to obtain access to Stage view without need to be authenticated again. The access token obtained for the app user when they first used your app can be used to allow them to use Stage view.

> [!TIP]

> > You can use also SSO API to authenticate app users in [task modules](../../../task-modules-and-cards/what-are-task-modules.md) that embed web content.

Reviewer: typo? Suggest adding an article before "SSO API" and checking word order.
Change to "You can also use the SSO API to authenticate ..."

To achieve SSO at runtime, configure your app to enable SSO for authenticating and authorizing app users.

## Enable SSO for a tab app

This section describes the tasks involved in implementing SSO for a tab app. These tasks are language- and framework-agnostic.

To enable SSO for a tab app:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; :::image type="content" source="../../../assets/images/authentication/teams-sso-tabs/enable-sso.png" alt-text="Steps to enable SSO for tab" border="false" lightbox="../../../assets/images/authentication/teams-sso-tabs/enable-sso.png":::

1. **Register with Azure AD**: Create an Azure AD app to generate an app ID and application ID URI. Configure redirect URI where Azure AD would send the access token in exchange for identity token for the current app user logged into Teams. For generating access token, you configure scopes and OBO flow.

> 2. **Configure code**: Update the code to handle access token, calling `getAuthToken()` when an app user accesses your tab app, and validating access token when it's received.

Reviewer: Suggest clarifying:

2. **Configure code**: Update the code to handle access token, calling `getAuthToken()` when an app user accesses your tab app, sending this token to your app's server code in the Authorization header, and validating the access token when it's received.

> 3. **Update Teams app manifest**: Update your Teams Client app manifest with the app ID generated on Azure AD and the application ID URI to ensure secure connection between Azure AD and your app.

Reviewer: Probably a nit but suggest rewording:

3. **Update Teams app manifest**: Update your Teams Client app manifest with the app ID and application ID URI generated on Azure AD to allow Microsoft Teams to request access tokens on behalf of your app.

## Best practices

Here's a list of best practices:

- **Call access token only when you need it**: Call `getAuthToken()` only when you need an access token. You can call it when an app user accesses your tab app, or for using a particular function that requires app user validation.
- **Don't store access token on client-side code**: Don’t cache or store the access token in your app's client-side code. Always call `getAuthToken()` when you need an access token. Teams Client caches the access token (or request a new one if it expires). This makes sure that there's no accidental leak of your token from your web app.
- **Use server-side code for Microsoft Graph calls**: Always use the server-side code to make Microsoft Graph calls, or other calls that require passing an access token. Never return the OBO token to the client to enable the client to make direct calls to Microsoft Graph. This helps protect the token from being intercepted or leaked. For more information on the proper protocol flow, see the [OAuth 2.0 protocol diagram](auth-flow-tab.md).

## Known limitations

> - Currently, SSO only supports OAuth 2.0 token. It doesn't support SAML token.

Reviewer: Suggest being explicit that this is Teams SSO not Azure:

> - Currently, Teams SSO only supports OAuth 2.0 tokens. It doesn't support SAML tokens.

Author: Thank you.
As we aren't using 'Teams SSO' as a term, I've fixed it as 'Currently, SSO in Teams supports only OAuth 2.0 token. It doesn't support SAML token.'

> - SSO in Teams works only with OAuth 2.0 protocol.

Reviewer: Isn't this redundant? Suggest deleting this point; it's an OAuth 2.0 token and the protocol is actually hidden within the getAuthToken() call.

> - SSO works only with apps that support Azure AD as identity provider. For implementing authentication with other OAuth identity providers, the OAuth flow needs to be implemented. For more information, see, [Enable authentication using third-party OAuth provider](auth-flow-tab.md) and [Configure third party OAuth authentication](auth-tab-aad.md).

Reviewer: Note that the authentication pop-up explained under the 1st link does not need to be OAuth! Ironically the authentication.authenticate() call does not authenticate users :-) -- it just shows a pop-up window and allows success/failure and data to be passed back to the tab. So any auth scheme that works in a browser would work here. Suggest changing this to:

- SSO works only with apps that support Azure AD as identity provider. For implementing authentication with other identity providers, the application must supply a pop-up web page to authenticate users. For more information, see, [Enable authentication using third-party OAuth provider](auth-flow-tab.md) and [Configure third party OAuth authentication](auth-tab-aad.md).

> - Multiple domains per app are not supported. For more information, see [LOB apps](tab-sso-register-aad.md#before-you-register-with-azure-ad).

Reviewer: Unless it's changed, it's also true that domains ending in "azurewebsites.net" are not supported, so if a developer deploys to Azure it won't work with the default settings. Loki said that was an Azure AD restriction. That was a long time ago - has it changed?

> - Tenant admin consent: A simple way of [consenting on behalf of an organization as a tenant admin](/azure/active-directory/develop/v2-permissions-and-consent.md#requesting-consent-for-an-entire-tenant) is by getting [consent from admin](https://login.microsoftonline.com/common/adminconsent?client_id=<AAD_App_ID).

Reviewer: The 2nd link is a sample link that isn't displayed in the markup :)
Suggest sending readers to the relevant docs article at https://docs.microsoft.com/en-us/azure/active-directory/manage-apps/grant-admin-consent

- Your app may have a bot, a tab, a message extension, or Adaptive Cards, and allows your app users to invoke task module from within a bot or a tab. Enabling SSO and using `getAuthToken()` will let your app users access a task module from within a bot app without having to sign in again. Teams can use the app user's Teams identity and obtain an access token from Azure AD.

- You can also use SSO to allow your app users to obtain access to Stage view without need to be authenticated again. The access token obtained for the app user when they first used your app can be used to allow them to use Stage view.

> [!TIP]
> You can also use the SSO API to authenticate app users in [task modules](../../../task-modules-and-cards/what-are-task-modules.md) that embed web content.

To achieve SSO at runtime, configure your app to enable SSO for authenticating and authorizing app users.

## Enable SSO for a tab app

This section describes the tasks involved in implementing SSO for a tab app. These tasks are language- and framework-agnostic.

To enable SSO for a tab app:

> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; :::image type="content" source="../../../assets/images/authentication/teams-sso-tabs/enable-sso.png" alt-text="Steps to enable SSO for tab" border="false" lightbox="../../../assets/images/authentication/teams-sso-tabs/enable-sso.png":::

Reviewer: Picture should not include OBO Flow; this is only used when calling a down-stream API such as Microsoft Graph

> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; :::image type="content" source="../../../assets/images/authentication/teams-sso-tabs/enable-sso.png" alt-text="Steps to enable SSO for tab" border="false" lightbox="../../../assets/images/authentication/teams-sso-tabs/enable-sso.png":::

Reviewer: NEW COMMENT:
The setting in the AAD registration is "Authorized client applications"; calling it "OBO Flow" may confuse users because "OBO" does not appear in the app registration settings.

> 1. **Register with Azure AD**: Create an Azure AD app to generate an app ID and application ID URI. Configure redirect URI where Azure AD would send the access token in exchange for identity token for the current app user logged into Teams. For generating access token, you configure scopes and OBO flow.

Reviewer: NEW COMMENT
Neither a redirect URI or OBO Flow are needed in this scenario. The "OBO Flow" is only used in the case of calling Microsoft Graph or some other downstream application; even in this case there is no "OBO Flow" setting in the app registration however an App Secret is needed for this.
The "Redirect URI" is only needed if an application implements OBO Flow and wants to handle consent for the downstream API, such as the app handling consent for User.Read Graph permissions. This is accomplished by using a pop-up (authentication.authenticate() call) that redirects the user to Azure AD where they can sign in and grant consent; then Azure AD redirects back to the app in the popup; the page that will handle this must be registered as the "Redirect URI"

---
title: Overview to authentication for tabs using SSO in Teams with Azure AD
description: Overview to SSO authentication in Teams and how to use it in tabs
ms.topic: how-to
ms.localizationpriority: medium
keywords: teams authentication tabs Microsoft Azure Active Directory (Azure AD)
---

> # Enable single sign-on in a tab application

Reviewer 2: Do we need to use the word application? or just tab would do?

> Single sign-on (SSO) lets a user sign-in only once to access an application or a web service. Once a user signs in, they can access the app without going through authentication again.

Reviewer 2: ```suggestion
Single sign-on (SSO) allows user to sign-in only once to access an application or a web service. After a user sign in, the user can access the app without going through authentication again.
```

With SSO in Teams, your app users have an added advantage. A user logged into Teams using their Microsoft or Microsoft 365 accounts doesn't need to sign in again to use your app. They can use your app within Teams environment based on access granted through Azure AD.

Here's what you'll learn in this section:

- SSO user experience in Teams, and why it's true SSO.
- SSO at runtime in Teams and how your app handles authenticating and authorizing the user.
- The steps you follow to enable SSO for your Teams app.

## SSO user experience in Teams

> Users sign in to Microsoft Teams using either their personal Microsoft account or their Microsoft 365 account. Take advantage of this, and use SSO to authenticate and authorize a Teams user.

Reviewer 2: Can we please rephrase to avoid usage of "their"

| 3 | Teams -> Azure AD | Teams requests Azure AD endpoint for the tab access token for the current user based on their Teams identity. |
| 4 | Azure AD -> Teams | Azure AD sends the tab access token to the Teams client. The token is a JSON Web Token (JWT), and it's validation works just like token validation in most standard OAuth flows. Teams will cache the token on your behalf so that future calls to getAccessToken simply return the cached token. |
| 5 | Teams -> Tab | Teams sends the tab access token to the tab as part of the result object returned by the `getAuthToken()` call. |
| 6 | Tab app | The tab app parses the token using JavaScript to extract required information, such as the user's email address. The token returned to the tab app is both an access token and an identity token. |

> [!IMPORTANT]
> The `getAuthToken()` is only valid for consenting to a limited set of user-level APIs that is email, profile, offline_access, and OpenId. It is not used for further Graph scopes such as `User.Read` or `Mail.Read`. For suggested workarounds, see [Extend your app with Microsoft Graph permissions](tab-sso-graph-api.md).

> [!NOTE]

> > Currently, SSO only supports OAuth 2.0 token and does not support SAML token.

Reviewer 3: Should we move this to Know limitation section?

Author: Moved the content to 'Known limitations'.

5. To authenticate a user in stage view.

The SSO API also works in [task modules](../../../task-modules-and-cards/what-are-task-modules.md) that embed web content.

> [!NOTE]
> Tabs are Teams-aware web pages. To enable SSO in a web page hosted inside a tab, add [Teams Javascript client SDK](/javascript/api/overview/msteams-client?view=msteams-client-js-latest&preserve-view=true), and call `microsoftTeams.initialize()`. Once you've initialized Microsoft Teams, call `microsoftTeams.getAuthToken()`, which will get the access token for your app.

To achieve Teams SSO, you must configure your app to enable SSO for authenticating and authorizing your users.

## Enable SSO for a tab app

This section describes the tasks involved in implementing SSO for a tab app. These tasks are language- and framework-agnostic.

To build a tab app that uses Teams SSO to authenticate users:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; :::image type="content" source="../../../assets/images/authentication/teams-sso-tabs/teams-sso-accesstoken.png" alt-text="Steps to enable SSO for tab" border="false":::

> 1. Register with Azure AD: Create an Azure AD app to generate an app ID and app ID URI You also configure redirect URI where Azure AD would send the access token in exchange for identity token for the current user logged into Teams.

Reviewer 3: ```suggestion
1. Register with Azure AD: Create an Azure AD app to generate an app ID and app ID URI. You also configure redirect URI where Azure AD would send the access token in exchange for identity token for the current user logged into Teams.
```

> 1. Register with Azure AD: Create an Azure AD app to generate an app ID and app ID URI You also configure redirect URI where Azure AD would send the access token in exchange for identity token for the current user logged into Teams.

Reviewer 3: I think making "Register with Azure AD" bold would look much better similar to "Best practices" section.

> 2. Configure code: Update the Teams client to handle access token, calling it when a user accesses your app and validating it when received.

Reviewer 3: ```suggestion
2. Configure code: Update the code to handle access token, calling it when a user accesses your app and validating it when received.
```

Author: Fixed the sentence as suggested.

3. Teams manifest: Update your Teams client manifest with the app ID generated on Azure AD and the app ID URI.

## Best practices

Here's a list of best practices:

- **Don’t cache the access token**: Never cache or store the access token in your app's client-side code. Always call `getAuthToken` when you need an access token. Teams Client caches the access token (or request a new one if it expires). This makes sure that there's no accidental leak of your token from your web-app.

## Known limitations

- SSO in teams only work with OAuth 2.0 protocol.  

> - SSO only works with Azure AD. To extend it to different OAuth Identity providers, the flow needs to be implemented.

Reviewer 3: > the flow needs to be implemented

It's not clear which flow we are referring here.  This should ideally be linked to some other part of the doc where we talk about solution.

Author: Added links to oAuth flow in Tabs  and Azure AD authentication pages.

> - Multiple domains per app is not supported. For this, please read about LOB apps.

Reviewer 3: > please read about LOB apps.

Need to provide link where we talk about LOB apps.

> - Tenant Admin Consent: A simple way of consenting on behalf of an organization as a tenant admin is to refer to `https://login.microsoftonline.com/common/adminconsent?client_id=<AAD_App_ID>`.

Reviewer 3: ```suggestion
- Tenant Admin Consent: A simple way of [consenting on behalf of an organization as a tenant admin](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#requesting-consent-for-an-entire-tenant ) is by getting consent from admin using this link: `https://login.microsoftonline.com/common/adminconsent?client_id=<AAD_App_ID>`.
```

Author: Added the links.

- Ask for consent using the Auth API: Another approach for getting Graph scopes is to present a consent dialog using our existing [web-based Azure AD authentication approach](~/tabs/how-to/authentication/auth-tab-aad.md#navigate-to-the-authorization-page-from-your-pop-up-page). This approach involves popping up an Azure AD consent dialog box.

    To ask for additional consent using the Auth API, follow these steps:

    1. The token retrieved using `getAuthToken()` must be exchanged server-side using Azure AD [on-behalf-of flow](/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) to get access to those other Graph APIs. Ensure you use the v2 Graph endpoint for this exchange.
    2. If the exchange fails, Azure AD returns an invalid grant exception. There are usually one of two error messages, `invalid_grant` or `interaction_required`.

>     3. When the exchange fails, you must ask for consent. Show some user interface (UI) asking the user to grant other consent. This UI must include a button that triggers an Azure AD consent dialog box using our [Azure AD authentication API](~/concepts/authentication/auth-silent-aad.md).

Reviewer 3: In same section we are referring two different auth types. 

This refers to simple auth implementation.
![image](https://user-images.githubusercontent.com/31851992/165985406-72fdfe0b-3460-435f-83f4-75cbd9165aaa.png)

This section refers to silent auth page.
![image](https://user-images.githubusercontent.com/31851992/165985556-1cef7178-0b6e-484b-b9c6-07612db39b08.png)
